packages/eslint-plugin/docs/rules/no-unnecessary-condition.md

```diff
@@ -97,6 +97,11 @@ If for some reason you cannot turn on `strictNullChecks`, but still want to use
 
 The main downside to using this rule is the need for type information.
 
+This rule has a known edge case of triggering on conditions that were modified within function calls (as side effects).
+It is due to limitations of TypeScript's type narrowing.
+We recommend you use an [inline ESLint disable comment](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1).
```

Comment:
Actually, to be exact you don't need disable to work around `#9998`. For example:

```js
let x = false;

const f = () => (x = true);
f();

if (x) console.log();
```

You just need an upcast `false as boolean`.